File: src/main/java/rocks/sakira/gentlefawn/register/Items.java
package rocks.sakira.gentlefawn.register;

import net.minecraft.item.*;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import rocks.sakira.gentlefawn.GentleFawn;
import rocks.sakira.gentlefawn.item.GentleFawnItemGroup;
import rocks.sakira.gentlefawn.item.SupplierSpawnEggItem;

public class Items {
    public static final DeferredRegister<Item> REGISTER = DeferredRegister.create(ForgeRegistries.ITEMS, GentleFawn.MOD_ID);

    private static final ItemGroup GROUP = new GentleFawnItemGroup("gentlefawn") {
        @Override
        public ItemStack createIcon() {
            return new ItemStack(DEER_SPAWN_EGG.get());
        }
    };

    public static final RegistryObject<SpawnEggItem> DEER_SPAWN_EGG = REGISTER.register(
            "deer_spawn_egg",

            () -> new SupplierSpawnEggItem(
                    Entities.DEER_ENTITY,
                    0xB49D8A,
                    0xF6F3EF,

                    (new Item.Properties())
                            .group(GROUP)
            )
    );

    public static final RegistryObject<Item> VENISON_ITEM = REGISTER.register(
            "venison",

            () -> new Item(
                    new Item.Properties()
                            .group(GROUP)
                            .food(Foods.BEEF)
            )
    );

    public static final RegistryObject<Item> COOKED_VENISON_ITEM = REGISTER.register(
            "cooked_venison",

            () -> new Item(
                    new Item.Properties()
                            .group(GROUP)
                            .food(Foods.COOKED_BEEF)
            )
    );

    public static final RegistryObject<Item> DOE_HEAD_ITEM = REGISTER.register(
            "doe_head",

            () -> new WallOrFloorItem(
                    Blocks.DOE_HEAD.get(),
                    Blocks.DOE_WALL_HEAD.get(),

                    new Item.Properties()
                            .group(GROUP)
                            .rarity(Rarity.UNCOMMON)
            )
    );

    public static final RegistryObject<Item> BUCK_HEAD_ITEM = REGISTER.register(
            "buck_head",

            () -> new WallOrFloorItem(
                    Blocks.BUCK_HEAD.get(),
                    Blocks.BUCK_WALL_HEAD.get(),

                    new Item.Properties()
                            .group(GROUP)
                            .rarity(Rarity.UNCOMMON)
            )
    );
}

File: src/main/kotlin/io/github/faecraft/gentlefawn/register/Items.kt
package io.github.faecraft.gentlefawn.register

import io.github.faecraft.gentlefawn.Common.MOD_ID
import net.fabricmc.fabric.api.client.itemgroup.FabricItemGroupBuilder
import net.fabricmc.fabric.api.item.v1.FabricItemSettings
import net.minecraft.block.Block
import net.minecraft.entity.EntityType
import net.minecraft.item.*
import net.minecraft.util.Identifier
import net.minecraft.util.Rarity
import net.minecraft.util.registry.Registry

object Items {
    val GROUP: ItemGroup = FabricItemGroupBuilder
        .create(Identifier(MOD_ID, MOD_ID))
        .icon { ItemStack(DOE_HEAD) }
        .build()

    val DEER_SPAWN_EGG = createSpawnEggItem(Entities.DEER, 0xB49D8A, 0xF6F3EF)

    val VENISON = createFoodItem(FoodComponents.BEEF)
    val COOKED_VENISON = createFoodItem(FoodComponents.COOKED_BEEF)

    val DOE_HEAD = createWallStandingBlockItem(Blocks.DOE_HEAD, Blocks.DOE_HEAD_WALL, Rarity.UNCOMMON)
    val BUCK_HEAD = createWallStandingBlockItem(Blocks.BUCK_HEAD, Blocks.BUCK_HEAD_WALL, Rarity.UNCOMMON)

    fun register() {
        register(DEER_SPAWN_EGG, "deer_spawn_egg")

        register(VENISON, "venison")
        register(COOKED_VENISON, "cooked_venison")

        register(DOE_HEAD, "doe_head")
        register(BUCK_HEAD, "buck_head")
    }

    private fun register(item: Item, id: String) {
        Registry.register(Registry.ITEM, Identifier(MOD_ID, id), item)
    }

    private fun createWallStandingBlockItem(blockStanding: Block, blockWall: Block, rarity: Rarity) = WallStandingBlockItem(
        blockStanding,
        blockWall,

        FabricItemSettings()
            .group(GROUP)
            .rarity(rarity)
    )

    private fun createFoodItem(food: FoodComponent) = Item(
        FabricItemSettings()
            .group(GROUP)
            .food(food)
    )

    private fun createSpawnEggItem(entity: EntityType<*>, primaryColor: Int, secondaryColor: Int) = SpawnEggItem(
        entity,
        primaryColor,
        secondaryColor,
        FabricItemSettings()
            .group(GROUP)
    )
}
